### PR TITLE
PR-Request review - Ruby group capstone - Catalog of my things

### PR DESCRIPTION
**Pull request by ivanmvh**
# Ruby group capstone - Catalog of my things

## Description

In this project, we will create a console app that will help we to keep a record of different types of things we own: books, music albums, movies, and games.
Everything will be based on the UML class diagram presented below. The data will be stored in JSON files but we will also prepare a database with tables structure analogical to our program's class structure.

## General requirements

- Make sure that there are no linter errors.
- Make sure that we used correct Gitflow.
- Make sure that we documented our work in a professional way.
- Follow our list of best practices for Ruby.

#### Logic

- Each class from the UML class diagram (./images/catalog_of_my_things.png) should be created in a separate .rb file.
- All associations (1-to-many relationships) and aggregations (parent-child relationships) between classes should be implemented.
- All properties visible in the diagram should be defined and set up in the constructor method. 
  Exception: properties for the 1-to-many relationships should NOT be set in the constructor method. 
  Instead, they should have a custom setter method created._
- All methods visible in the diagram should be implemented:
    - add_item method (in all classes that have that method)
        - should take an instance of the Item class as an input.
        - should add the input item to the collection of items.
        - should add self as a property of the item object (by using the correct setter from the item object).
    - can_be_archived?() in the Item class
        - should return true if published_date is older than 10 years.
        - otherwise, it should return false.
    - move_to_archive() in the Item class
        - should reuse can_be_archived?() method.
        - should change the archived property to true if the result of the can_be_archived?() method is true.
        - should do nothing if the result of the can_be_archived?() method is false.
    - can_be_archived?() in the Book class
        - should override the method from the parent class.
        - should return true if parent's method returns true OR if cover_state equals to "bad".
        - otherwise, it should return false.
    - can_be_archived?() in the MusicAlbum class
        - should override the method from the parent class.
        - should return true if parent's method returns true AND if on_spotify equals true.
        - otherwise, it should return false.
    - can_be_archived?() in the Movie class
        - should override the method from the parent class.
        - should return true if parent's method returns true OR if silent equals true.
        - otherwise, it should return false.
    - can_be_archived?() in the Game class
        - should override the method from the parent class.
        - should return true if parent's method returns true AND if last_played_at is older than 2 years.
        - otherwise, it should return false.
- Add unit tests for all implemented methods.

#### User interface

- Create a main.rb file that will serve as our console app entry-point.
- our console app, at the start, should:
    - Present the user with a list of options to perform.
    - Let users choose an option.
    - If needed, ask for parameters for the option.
    - Have a way to quit the app.
- The following options should be available:
    - List all books
    - List all music albums
    - List all movies
    - List of games
    - List all genres (e.g 'Comedy', 'Thriller')
    - List all labels (e.g. 'Gift', 'New')
    - List all authors (e.g. 'Stephen King')
    - List all sources (e.g. 'From a friend', 'Online shop')
    - Add a book
    - Add a music album
    - Add a movie
    - Add a game
 - All data should be preserved by saving collections in .json files.

### Database
- Create a schema.sql file with tables that will be analogical to the structure of the classes in our app. As we cannot implement inheritance in the database tables - add all properties and associations from the parent Item class as table columns to all tables based on the child classes. Thanks to that it can be used to store data in the future.

### Project documentation

Once we have finished the development of the project, we should record a video presenting the features of the project we built. It is a video with a **maximum length of 5 minutes**. The content of the video should include:

- A description of the project.
- A demo of the project features.
- we should also highlight some interesting piece of code or something we built that we are very proud of.
- we all should appear in the video and talk about the project.

## Additional requirements

*These are all optional, but if we're interested in exploring this topic further, feel free to implement them. Any exploration here should be done outside program time.*

- Think about other options we might implement in our console app, e.g:
    - remove a selected book
    - add genre to selected book
    - ...

@ivanmvh

### CODE REVIEW

File: spec/musica_album_spec.rb [hunk 1]
```diff
@@ -0,0 +1,12 @@
...
+require 'date'
+
+describe 'Music Album class test cases.' do
+  context 'Three parameters will be assigned.' do
+    it 'Will return an object' do
+      album = MusicAlbum.new(publish_date: '2022-04-10', on_spotify: true)
+      expect(album).to be_an_instance_of MusicAlbum
+    end
+  end
+end
```
Comment (VICTOR-CHUKWU) [inline spec/musica_album_spec.rb:12]: - [ ] kindly add test for the method can_be_archived
```
    def can_be_archived?
      super && @on_spotify == true
    end
```

File: lib/item.rb [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+
+    author.items.push(self) unless author.items.include?(self)
+  end
+
+  private
+
+  def can_be_archived?
+    (Date.today - Date.parse(@publish_date)).to_i / 365 > 10
+  end
+end
```
Comment (VICTOR-CHUKWU) [inline lib/item.rb:31]: - [ ] in your spec folder I did not see any test for item class, kindly test methods like `move_to_archive` and  `can_be_archived?`

File: README.md [hunk 3]
```diff
@@ -1 +1,251 @@
...
+Database
+
+    Create a schema.sql file with tables that will be analogical to the structure of the classes in your app. As you cannot implement inheritance in the database tables - add all properties and associations from the parent Item class as table columns to all tables based on the child classes. Thanks to that it can be used to store data in the future.
+
+
+<p align="right">(<a href="#readme-top">back to top</a>)</p>
+
+<!-- LIVE DEMO -->
+
+## 🚀 Project Presentation <a name="live-demo"></a>
```
Comment (VICTOR-CHUKWU) [inline README.md:125]: - [ ] the video presentation should be of 5mins maximum length, you can trim it to fit

**Review by VICTOR-CHUKWU — CHANGES_REQUESTED** [review]
Hi TEAM, 
## STATUS: CHANGES REQUIRED :recycle: 
Good job so far!
There are some issues that you still need to work on to prepare your project for the final evaluation but you are almost there!


## Suggested changes
_Check the comments under the review._

You can use as many of my suggestions as you want. If there is anything you would like to skip - feel free to do that. However, I strongly recommend you to take them into account as they can make your code better._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

**Review by Yerimah — APPROVED** [review]
# STATUS: APPROVED 🔥 

Hi @HunterX-7 @cilfonegabriel @ivanmvh, 👋 

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 

68747470733a2f2f36342e6d656469612e74756d626c722e636f6d2f38376530623634363963393537626638393663646563316361623634366163662f74756d626c725f696e6c696e655f706b316264616b434370317270637239615f3530302e676966

## Highlights
- Descriptive PR summary ✔️ 
- No linter errors ✔️ 
- GitHub and Gitflow ✔️ 

## Optional suggestions

- [ ] I would kindly suggest that you add `title ` and `genre` to the attr_accessor method in the music album file. So that users can add the title of the album as well as genre. This will make your work better. 

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**